v2/pkg/engine/datasource/grpc_datasource/fetch.go

```diff
@@ -0,0 +1,144 @@
+package grpcdatasource
+
+import (
+	"fmt"
+	"sync"
+)
+
+type FetchItem struct {
+	ID               int
+	Plan             *RPCCall
+	ServiceCall      *ServiceCall
+	DependentFetches []int
+}
+
+type DependencyGraph struct {
+	mu      sync.Mutex
+	fetches []FetchItem
+	// nodes is a list of lists of dependent calls.
+	// Each node index corresponds to a call index in the execution plan
+	// and the list contains the corresponding dependent calls indices.
+	nodes [][]int
+}
+
+func NewDependencyGraph(executionPlan *RPCExecutionPlan) *DependencyGraph {
+	graph := &DependencyGraph{
+		nodes:   make([][]int, len(executionPlan.Calls)),
+		fetches: make([]FetchItem, len(executionPlan.Calls)),
+	}
+
+	for index, call := range executionPlan.Calls {
+		graph.nodes[index] = call.DependentCalls
+		graph.fetches[index] = FetchItem{
+			ID:               index,
+			Plan:             &call,
+			ServiceCall:      nil,
+			DependentFetches: call.DependentCalls,
+		}
+	}
+
+	return graph
+}
+
+func (g *DependencyGraph) TopologicalSortResolve(resolver func(nodes []FetchItem) error) error {
```

Comment:
This is heavy function with recursive traversal. Could you comment it what it does?

As I have understood from the code, it arranges calls into levels (chunks) so calls on each level can be executed in parallel. Level 0 does not depends on anything and Level 1 depends on level 0, and so on.